Add initial surface ocean BGC capability

bradyrx opened the pull request:
This PR adds a new task `climatologyMapBGC` which plots a variety of surface ocean BGC variables:

>  PO4, NO3, SiO3, CO2 flux, pCO2, pH, DIC, ALK, O2, Chl

An example BGC-only output page is [hosted here](http://storm.colorado.edu/~ribr5703/bgc_spinup/). 

Observations are from GLODAP, WOA, a neural network product, and SeaWiFS. Post-processed obs can be found under `/usr/projects/climate/SHARED_CLIMATE/observations/BGC/` on IC. This PR also contains a script for downloading and processing raw BGC observations.

Example config files for testing can be found under:
*  configs/lanl/config.BGCphaeo4.testrun : 60to30 (no pCO2)
* configs/lanl/config.GMPAS-OECO-ODMS-IAF.oRRS30to10v3 : 30to10

milenaveneziani commented:
This looks great @bradyrx, thanks! 
(btw, I'm sorry we haven't met in person yet. Hopefully at the next meeting. Where are you sitting?).

Do we know if the observations can be publicly released or if there is a specific policy to do that?

milenaveneziani commented:
@maltrud: I added you as a reviewer because of the BGC component. Should I add Shanlin too?

bradyrx commented:
@milenaveneziani, observations are all publicly released, most with a request to include their citation when being used. I've added that info to the `observational_datasets.xml` file.

I'm sitting in cube 3 in building 200. I can introduce myself at the next meeting, though.

xylar commented on mpas_analysis/config.default:
```diff
@@ -266,6 +266,25 @@ sshSubdirectory = SSH
+sstClimatologyEndYear = 1900
+# alternative values for present day
+#sstClimatologyStartYear = 1990
+#sstClimatologyEndYear = 2011
```
@bradyrx, I think this is a glitch from a rebase.  This stuff got moved to a different section of the config file and should be removed here.

xylar commented on mpas_analysis/ocean/climatology_map_bgc.py:
```diff
@@ -0,0 +1,356 @@
+# Copyright (c) 2017,  Los Alamos National Security, LLC (LANS)
```
@bradyrx, if you don't mind, could you run a PEP8 style checker or alternatively open this file in spyder (and, I think, turn on PEP8 checking) or another IDE that includes PEP8 style checking? We try to stick pretty closely with that standard. Mostly what I'm seeing is lines with trailing whitespace and places without the expected number of line breaks. If you'd rather not deal with this, that's totally fine and I can clean it up later.

bradyrx replied:
@xylar fixed whitespace and other PEP8 issues.

xylar commented on mpas_analysis/ocean/climatology_map_bgc.py:
```diff
@@ -0,0 +1,356 @@
+        
+        # Clarify that the user is doing preindustrial vs. modern
+        preindustrial = config.getboolean(sectionName, 'preindustrial')
+        if preindustrial:
```
@bradyrx, could you create a `setup_and_check` function below and move this bit (including the print statements below) into that function?  That way, this check will only happen and the message will only be printed if the user has requested this specific analysis.  I'm seeing this message even when I do, for example, `./run_mpas_analysis --list`.  In general, the only print statements in the `__init__` function of a task should be related to catastrophic errors.

xylar commented:
@bradyrx, this all looks very good to me.  Other than my few comments above, I think this is ready to merge once we've created the v1.0 tag (which is waiting on licensing information from LANL and LLNL).

I haven't actually tried the BGC runs on LANL IC yet, so I'll also give those a try once I'm back at the lab next week.

bradyrx commented:
@xylar, alright I fixed the PEP8 issues, moved the preindustrial check and print statement down to `setup_and_check`, and fixed the rebase issues in config.default. I think that should be everything, pending a full LANL IC check by you and the licensing. Thanks!

xylar commented:
Great, thanks!

xylar commented:
## Testing

I am performing a "do no harm" test with a QU240wLI on my laptop right now.  This run doesn't have any BGC so it's just a test to make sure the tasks that existed before this PR still work as they should.  I'll report back once the test completes, but so far, so good.

bradyrx commented:
@milenaveneziani Not to my knowledge. This implements surface climatology maps for high-priority BGC variables. I imagine future updates could include transects, depth profiles, etc.

milenaveneziani commented:
Great. Thanks @bradyrx!

xylar commented:
@milenaveneziani and @maltrud, do you want to review this?  @maltrud, that would probably mean making sure the output is what you wanted.  @milenaveneziani, I'm not sure, maybe it would just mean a quick look through the code in your case to make sure nothing jumps out at you.  You're quite good at picking up on strangeness that way.

xylar commented:
It only took @milenaveneziani 4 seconds to read my message, read the code, decide it was fine, and approve it ;-)

bradyrx commented:
@xylar @milenaveneziani  -- BGC functions properly after rebasing with a simple test case ([link](http://storm.colorado.edu/~ribr5703/rebase_test/)).

xylar commented:
That's great, how exciting!  I'm running a few tests on my laptop right now.  I'll let you know how it goes.

xylar commented:
One thing that's missing is `docs/tasks/climatologyMapBGC.rst`, the basic documentation for the task.  My feeling is that this is something that I can easily add as part of the refactoring I'm about to do.

xylar commented:
This branch passed 2 tests on my laptop with a QU240wLI run and default settings (i.e. no BGC).  The third is still running but it's far enough along that I don't anticipate any trouble.  I'm ready to merge.  Nice work, @bradyrx!

xylar commented:
@bradyrx, you can now delete your remote (and local) branches if you want.  Let me know if you need help with that.
